### ISSUE TITLE
Various code size optimisations can be made

### ISSUE THREAD

**Issue by chris-morgan** [lukeed/navaid#2]
You seem to care about code size (since you describe it so very early on), but there was a fair bit left on the table. I’m just dumping this here, it should be semantically equivalent, except for not using non-standard events, and fixing the pushState/replaceState arguments; but I have no inclination to follow anything up on it. If one truly wants to minimise size, regexparam would need to be inlined, and its objects turned into arrays. Anyway, you may do what you will with this, down to and including nothing. Enjoy!

```js
import convert from 'regexparam';

export default (base, on404) => {
	let rgx, routes=[], handlers={}, $, hist = history;

	let fmt = uri => {
		if (!uri) return uri;
		uri = '/' + uri.replace(/^\/|\/$/g, '');
		return rgx ? rgx.test(uri) && (uri.replace(rgx, '') || '/') : uri;
	};

	if ((base=fmt(base)) === '/') base = '';
	if (base) rgx = new RegExp('^/?' + base.substring(1) + '(?=/|$)', 'i');

	return $ = {
		format: fmt, 

		route(uri, replace) {
			hist[replace ? 'replaceState' : 'pushState']('', '', base + uri);
		},

		on(pat, fn) {
			handlers[pat] = fn;
			let o = convert(pat);
			o.route = pat;
			routes.push(o);
			return $;
		},

		run(uri) {
			uri = fmt(uri || location.pathname);
			for(var i=0, len=routes.length;i<len;i++){
				let obj=routes[i], arr=obj.pattern.exec(uri);
				if (arr) {
					let keys=obj.keys, params={};
					// Ain’t variable reuse fun?
					i=keys.length;
					while (i--) params[keys[i]]=arr[i] || null;
					handlers[obj.route](params); // todo loop?
					return $;
				}
			}
			if (uri && on404) {
				on404(uri);
			}
			return $;
		},

		listen() {
			let off = removeEventListener;
			let on = addEventListener;
			let run = () => $.run();
			let click = e => {
				let x = e.target.closest('a');
				if (!x || !x.href || x.target) return;
				if (e.ctrlKey || e.metaKey || e.altKey || e.shiftKey || e.button) return;
				if (x = fmt(x.getAttribute('href'))) {
					e.preventDefault();
					$.route(x);
				}
			};

			let pushState = hist.pushState;
			let replaceState = hist.replaceState;

			hist.pushState = (state, title, uri) => {
				pushState.call(hist, state, title, uri);
				run();
			};
			hist.replaceState = (state, title, uri) => {
				replaceState.call(hist, state, title, uri);
				run();
			};

			on('popstate', run);
			on('click', click);

			$.unlisten = () => {
				hist.pushState = pushState;
				hist.replaceState = replaceState;
				off('popstate', run);
				off('click', click);
			};

			return $.run();
		},
	};
};
```

Feeding this through Terser for minification (without inlining the dependency), normal/gzipped: before, 1274/671 bytes; after, 1061/608 bytes. Roughly 16%/8% savings.

There are various further optimisations that can be performed to shave fifty or more bytes off the minified result, but at the cost of a few bytes gzipped—turning two or three similar things into a function that you call two or three times can often save of ungzipped size, at the cost of gzipped size because the repetition compresses so well.

P.S. I strongly dislike the history.{pushState, replaceState} replacement. $.route() and the <a> interception should be the only way to do that. Oh yeah, a couple of things that are missing here: support for imagemaps, and absolute URLs—for the latter, I suggest using HTMLHyperlinkElementUtils.href, which is a full URL, and prefix-matching on a full URL base (which will probably be an origin plus a slash). I’m done now.

**Comment by lukeed** [comment]
Hey, thank you~! 🙌 

I played with it a bit – your 671/608 doesn't include `regexparam`, but my numbers will here.

As presented (and with the latest version of `regexparam` included), this code is 738B after Terser. 

However, this isn't ES5... I know it's lame to still be shipping ES5 in 2019, but unfortunately it's what still needs to be done given the standard/CRA-like tools that don't transpile `node_modules` by default. I think this will be the year that changes that, but I don't want to be fighting it/catching the brunt of it in the meantime.

When expanding your code to ES5, it grows to 751B. 

I liked your `$.run` improvements and was surprised I had left that extra `find()` loop 😅 I added that in & upgraded the repo to use Terser, and the current size is 788B.

That leaves the `wrap`/event emitter in that 37B, which I'm in favor of & actually consider a feature of Navaid. It normalizes the history events across browsers & actually allows one to respond specifically to replace, push, and pop _individually_ and outside of the router context. This has come in handy on so many occasions that I can't even count anymore
